File: 59-spiral-matrix-ii/59-spiral-matrix-ii.py
class Solution(object):
    def generateMatrix(self, n):
        """
        :type n: int
        :rtype: List[List[int]]
        """
        direction = [(-1,0),(0,-1),(1,0),(0,1)]
        stack = [m for m in range(1,n*n+1)]
        matrix = [[0]*n for _ in range(n)]
        x, y =0, 0
        for i in range(n*n):
                matrix[x][y] = stack[i]
                newx, newy = x + direction[-1][0], y + direction[-1][1]
                if 0 <= newx < n and 0 <= newy < n and matrix[newx][newy] == 0:
                    x, y = newx, newy
                else:
                    direction = [direction.pop()] + direction
                    x, y = x + direction[-1][0], y + direction[-1][1]
        return matrix
                    
                    
                
            
        
        

File: 59-spiral-matrix-ii/59-spiral-matrix-ii.java
class Solution {
    public int[][] generateMatrix(int n) {
        int[][] res = new int[n][n];
        int[][] directions = {{0,1},{1,0},{0,-1},{-1,0}};
        int x = 0, y = 0;
        res[x][y] = 1;
        int j = 0;
        for(int i = 2; i < n*n + 1; i++){
            while(true){
            int newx = directions[j%4][0] + x, newy = directions[j%4][1] + y;
            if(newx < n && newx >= 0 && newy < n && newy >= 0 && res[newx][newy] == 0){
                res[newx][newy] = i;
                x = newx;
                y = newy;
                break;
            }else{
                j++;
            }
        }
        
        }
        return res;
        
        
    }
}